### PR TITLE
[Fix #6617] Prevent traversal error on infinite ranges

### LINKED ISSUE #6617: 0.61.1: Rubocop crashes with infinite ranges
Test files:

```
$ cat .rubocop.yml 

AllCops:
  TargetRubyVersion: 2.6

$ cat test.rb

def foo
  bar[11..]
end
```

--------

## Expected behavior

Rubocop should work normally.

## Actual behavior

```
Inspecting 1 file


0 files inspected, no offenses detected
undefined method `type' for nil:NilClass
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:50:in `block in on_irange'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:50:in `each'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:50:in `on_irange'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/commissioner.rb:35:in `block (2 levels) in <class:Commissioner>'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:107:in `block in on_send'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:104:in `each'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:104:in `each_with_index'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:104:in `on_send'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/commissioner.rb:35:in `block (2 levels) in <class:Commissioner>'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:100:in `on_def'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/commissioner.rb:35:in `block (2 levels) in <class:Commissioner>'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/ast/traversal.rb:13:in `walk'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/commissioner.rb:46:in `investigate'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/team.rb:116:in `investigate'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/team.rb:104:in `offenses'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cop/team.rb:44:in `inspect_file'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:280:in `inspect_file'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:227:in `block in do_inspection_loop'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:259:in `block in iterate_until_no_changes'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:252:in `loop'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:252:in `iterate_until_no_changes'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:223:in `do_inspection_loop'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:126:in `block in file_offenses'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:144:in `file_offense_cache'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:124:in `file_offenses'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:112:in `process_file'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:89:in `block in each_inspected_file'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:86:in `each'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:86:in `reduce'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:86:in `each_inspected_file'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:76:in `inspect_files'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/runner.rb:48:in `run'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cli.rb:150:in `execute_runner'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cli.rb:77:in `execute_runners'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/lib/rubocop/cli.rb:42:in `run'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/exe/rubocop:13:in `block in <top (required)>'
/Users/jablko/.rvm/rubies/ruby-2.6.0/lib/ruby/2.6.0/benchmark.rb:308:in `realtime'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/gems/rubocop-0.61.1/exe/rubocop:12:in `<top (required)>'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/bin/rubocop:23:in `load'
/Users/jablko/.rvm/gems/ruby-2.6.0@r5/bin/rubocop:23:in `<main>'
```

## Steps to reproduce the problem

After creating two files mentioned in the beginning, simply run `rubocop`.

## RuboCop version

```
$ rubocop -V
0.61.1 (using Parser 2.5.3.0, running on ruby 2.6.0 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by Drenmi**
With the addition of infinite ranges, the `irange` and `erange` node types can now have one node type child and one `nil` type child. Our code made the assumption that there would always be two node type children, causing a `NoMethodError` on traversal of infinite ranges. This change fixes that.

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [ ] Added tests.
* [X] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [X] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
I guess we also need some regression test for this.

**Comment by Drenmi** [comment]
> I guess we also need some regression test for this.

Was just looking at where to add this. 🤔

File: lib/rubocop/ast/traversal.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def on_#{type}(node)
       MANY_CHILD_NODES.each do |type|
         module_eval(<<-RUBY, __FILE__, __LINE__ + 1)
           def on_#{type}(node)
-            node.children.each { |child| send(:"on_\#{child.type}", child) }
+            node.children.compact.each { |child| send(:"on_\#{child.type}", child) }
```
Comment (bquorning) [inline lib/rubocop/ast/traversal.rb]: `compact` isn’t used anywhere else in this file. But a few methods have `if child` instead – e.g. `#on_case` which is now (functionally) exactly identical to `#on_irange` and the other `MANY_CHILD_NODES` methods.

I guess what I’m suggesting is: Couldn’t we add `case`, `rescue`, `resbody`, `ensure`, `for` and `when` to `MANY_CHILD_NODES`?

Comment (bquorning) [reply lib/rubocop/ast/traversal.rb]: `const` could be added to `ONE_CHILD_NODE`, and `on_if` looks identical to `on_class`. I think this file needs a refactoring…

Comment (Drenmi) [reply lib/rubocop/ast/traversal.rb]: @bquorning Great point. I changed the approach to alias `#on_case` instead.

Seems like you have a pretty good idea for how to make this code better. Will leave the additional improvements to you. 🙏

Comment (bquorning) [reply lib/rubocop/ast/traversal.rb]: I never noticed this piece of code before yesterday 😆 But when I did, I started studying and rewriting it.

**Comment by Drenmi** [comment]
@bbatsov Added a `RangeNode` extension that serves as a regression test for this as well. 🙇

**Comment by bbatsov** [comment]
The build is failing.

**Comment by bbatsov** [comment]
Roger that! Thanks for the update.

**Pull request merged**